### PR TITLE
Remove WorldCell

### PR DESCRIPTION
**Pull request by alice-i-cecile**
# Objective

- `WorldCell` is not fully featured: see #1555.
- As a result, it does not fully behave like a `Cell` (a runtime borrow-checked API to split borrows arbitrarily).
- Making it fully-featured is a questionable idea: see discussion in https://github.com/bevyengine/rfcs/pull/42 on how this promotes very panic-prone, fragile and hard-to-reason about code that is easy to write.

## Solution

- Remove `WorldCell` completely, rather than keeping a half-working implementation around.
- Users should use `SystemState` or `resource_scope` as needed.
- As other clear use cases arise that cannot be comfortably tackled with the existing API, add additional, less error-prone methods on `World`.

## Notes

In the places where this was used:
- several places were exclusive systems that did not need to be exclusive
- a couple usages were trivially replaced with a resource_scope
- other usages were trivially replaced with a `SystemState`
- some were apparently completely unnecessary (assorted pipeline `FromWorld` impls)
- one (`winit_runner_with`) was straightforward but rather verbose, and reflective of generally not-great code quality (it should be reusable but isn't, and is rather unstructured)

### CODE REVIEW

**Comment by alice-i-cecile** [comment]
@bevyengine/ecs-team; this is fundamentally a design decision. RFC #42 seems fairly doomed (a decision I support), but if it's doomed we should close it and remove the half-working `WorldCell` API.

As a point in favor of this removal, this API was *never* used internally.

**Comment by cart** [comment]
> As a point in favor of this removal, this API was never used internally.

This isn't true :)

image

**Comment by alice-i-cecile** [comment]
Ah, my search missed it. Let me try fixing those...

**Comment by cart** [comment]
Pretty sure they can all be replaced with SystemState or resource_scope.
What WorldCell really enables is "arbitrary nesting of non-conflicting contexts with arbitrary ad-hoc borrows". Bevy doesn't use that pattern internally except for the RenderGraph resource_scope, but I can see why some users might want it. _We_ might want it if we want to support Entitas-style event handlers.

**Comment by alice-i-cecile** [comment]
> What WorldCell really enables is "arbitrary nesting of non-conflicting contexts with arbitrary ad-hoc borrows"

I agree.

This is a *very* risky pattern, and non-trivial uses of this are likely to lead to complex runtime panics whenever surrounding code is changed. IMO the ideas discussed in #2192 are a much better way to accomplish the same goals.

**Comment by alice-i-cecile** [comment]
It's a bit frustrating not to have a better solution for #2192 yet; the extremely complex system state constructed in `winit_runner` would particularly benefit from the caching and complexity management. Looking at the code, I actually think it may be cleanest to just create / store / run a little schedule on the App to push the information from `winit` into the various event resources. This would also make `winit_runner` much more reusable / hackable.

Overall, that refactor was quite painless, even for extremely complex uses of `WorldCell`. And now the system state can be cached and the borrow checker rules are upheld at compile time.

**Comment by TheRawMeatball** [comment]
> We might want it if we want to support Entitas-style event handlers.

I personally think requiring &mut World and adding hooks would be a better way to support this.

File: crates/bevy_winit/src/lib.rs [hunk 1]
```diff
@@ -263,9 +264,52 @@ pub fn winit_runner_with(mut app: App) {
                 window_id: winit_window_id,
                 ..
             } => {
-                let world = app.world.cell();
-                let winit_windows = world.get_resource_mut::<WinitWindows>().unwrap();
-                let mut windows = world.get_resource_mut::<Windows>().unwrap();
+                let mut system_state: SystemState<(
```
Comment (TheRawMeatball) [inline crates/bevy_winit/src/lib.rs:267]: I think i agree with clippy on this one, this type is horrendeous. I'd recommend making a custom struct and using `#[derive(SystemParam)]` so it's neater.

**Review by TheRawMeatball — COMMENTED** [review]
I agree with the idea, but the winit runner can be done better imo.

**Review by TheRawMeatball — APPROVED** [review]

**Comment by aevyrie** [comment]
Please ping me when this is ready to merge, assuming it happens after #3974. I want to double check that everything still behaves as expected. 🙂

File: crates/bevy_pbr/src/render/light.rs [hunk 2]
```diff
@@ -162,52 +162,53 @@ pub struct ShadowPipeline {
...
-                mipmap_filter: FilterMode::Nearest,
-                compare: Some(CompareFunction::GreaterEqual),
-                ..Default::default()
-            }),
-        }
+                ],
+                label: Some("shadow_view_layout"),
+            });
+
+            let mesh_pipeline = world.get_resource::<MeshPipeline>().unwrap();
```
Comment (aevyrie) [inline crates/bevy_pbr/src/render/light.rs:185]:
```suggestion
            let mesh_pipeline = world.resource::<MeshPipeline>();
```

File: crates/bevy_render/src/lib.rs [hunk 3]
```diff
@@ -121,8 +121,10 @@ impl Plugin for RenderPlugin {
         if let Some(backends) = options.backends {
             let instance = wgpu::Instance::new(backends);
             let surface = {
-                let world = app.world.cell();
-                let windows = world.get_resource_mut::<bevy_window::Windows>().unwrap();
+                let windows = app
+                    .world
+                    .get_resource_mut::<bevy_window::Windows>()
+                    .unwrap();
```
Comment (aevyrie) [inline crates/bevy_render/src/lib.rs:127]:
```suggestion
                    .resource_mut::<bevy_window::Windows>();
```

**Review by aevyrie — CHANGES_REQUESTED** [review]

**Comment by aevyrie** [comment]
Got it! We'll just have to be mindful if/when this and #3974 are merged that we don't accidentally break anything in the merging process. We don't have an effective way to automatically test the window updating just yet.

**Comment by alice-i-cecile** [comment]
Closing in favor of the dramatically less controversial #5109. I still think this should be done, but the code-quality changes should be ported over ASAP anyways.

**Comment by james7132** [comment]
Our internal uses of `World::cell` outside of tests is now at 0. It might be a good time to revisit this now that we have `UnsafeWorldCell` and `DeferredWorld`.

**Pull request closed without merging**